### ISSUE TITLE
Customer tags in snmp.inputs.field

### ISSUE THREAD

**Issue by abitrusty** [influxdata/telegraf#7358]
Hi,

I've read a few posts that say this can't be done but they are all 3-4 years old so thought I'd jsut ask again.

I'm looking for a way to put custom tags under each field. We have a shared device that has multiple customers on it and I need a way of polling multiple OIDs with the same field but tag them differently. I can get around this by creating multiple inputs but this will result in my polling the device 50 times ever minute which will most likely cause me problems

Is this possible yet with telegraf 1.14?
What I'd ideally like to work is

[[inputs.snmp]]
 agents = [ "X.X.X.X" ]

[[inputs.snmp.field]]
    name = "outbound"
    tag  = "Customer 1"
    oid = ".1.3.6.1.4.1.9.9.63.1.3.8.6.2.2.1"

[[inputs.snmp.field]]
    name = "outbound"
    tag = "Customer 2"
    oid = ".1.3.6.1.4.1.9.9.63.1.3.8.6.2.2.2"

[[inputs.snmp.field]]
    name = "outbound"
    tag = "Customer 3"
    oid = ".1.3.6.1.4.1.9.9.63.1.3.8.6.2.2.3"

Thanks

**Comment by dgarros** [comment]
@abitrusty have you consider using a processor for that and add tags outside of the snmp plugin ? 
I've been using `processors.regex` with good success 
here is an example
```
[[processors.regex]]
  namepass = ["interface"]

  [[processors.regex.tags]]
    key = "name"
    pattern = "^Ethernet0$"
    replacement = "customerA"
    result_key = "customer_name"

  [[processors.regex.tags]]
    key = "name"
    pattern = "^Ethernet1$"
    replacement = "customerB"
    result_key = "customer_name"
```

otherwise if the name of the customer is available in SNMP there is a way to do that too using is_tag=True

```
   [[inputs.snmp.field]]
     name = "device"
     oid = "RFC1213-MIB::sysName.0"
     is_tag = true
```

**Comment by abitrusty** [comment]
Interesting. Didn't know about the regex. The second option I had throught of but not possislbe I'm afraid.

Had a quick read but I'm not clear on how this works.
Do i create another .conf file in telegraph directory ?  
Does it automatically run the regex on every snmp.conf file before it inserts to influx?
How would I just add a tag into the above fields where the only difference is the OID field?  Would it be like below?

[[processors.regex.tags]]
    key = "outbound"
    pattern = ".1.3.6.1.4.1.9.9.63.1.3.8.6.2.2.1"
    result_key = "Customer 1"

[[processors.regex.tags]]
    key = "outbound"
    pattern = ".1.3.6.1.4.1.9.9.63.1.3.8.6.2.2.2"
    result_key = "Customer 2"

[[processors.regex.tags]]
    key = "outbound"
    pattern = ".1.3.6.1.4.1.9.9.63.1.3.8.6.2.2.3"
    result_key = "Customer 3"


Many thanks in advance

**Comment by danielnelson** [comment]
Before getting into processors, is there any chance the data you are collecting is in a SNMP Table?

**Comment by abitrusty** [comment]
Hi Daniel, I was looking at tables but don't really understand what I can do with them. The majority of table example I see is the ifTable for Cisco switches etc

What I'm doing is getting active calls from a Cisco voice gateway. Cisco voice gateways use "dial-peers" to route calls and each customer has a specific dialpeer

In the diagram below I have cirlced the OID in question which looks like its in a table- I jsut add the dialpeer tag number to the end of that to get value of active calls

ie .1.3.6.1.4.1.9.9.63.1.3.8.4.1.1.100  (gets value of dialpeer 100 - CustomerA)
   .1.3.6.1.4.1.9.9.63.1.3.8.4.1.1.200   (gets value of dialpeer 200 - CustomerB)
![image](https://user-images.githubusercontent.com/54104859/79691739-286aa780-8259-11ea-86b8-9b60c1de6aa6.png)

So basically I want to be able to poll the one gateway but tag different customer names

[[inputs.snmp]]
agents = [ "X.X.X.X" ]

[[inputs.snmp.field]]
name = "inbound"
tag = "Customer A"
oid = ".1.3.6.1.4.1.9.9.63.1.3.8.4.1.1.100"

[[inputs.snmp.field]]
name = "outbound"
tag = "Customer B"
oid = ".1.3.6.1.4.1.9.9.63.1.3.8.4.1.1.200"

If I can do this with tables that would be great. Any help/advice is greatly appreciated too

**Comment by abitrusty** [comment]
So I've managed to pull down the table I need but I've been playing with tags and cannot get it to work how I need. Any advice on how I need to configure to do one poll of device 10.101.13.61 but have custom tags for each OID below?

![image](https://user-images.githubusercontent.com/54104859/79776240-5d482e80-832d-11ea-94bd-2818aa0313a6.png)

**Comment by danielnelson** [comment]
Here are two versions of how I think the config should be setup, of course I can't exactly test as I don't have these variables.

With the MIB installed on the Telegraf system:
```toml
[[inputs.snmp]]
  ## ... snip ...

  [[inputs.snmp.table]]
    oid = "CISCO-VOICE-DIAL-CONTROL-MIB::cvCallVolPeerTable"
```

Without the MIB installed:
```toml
[[inputs.snmp]]
  ## ... snip ...

  [[inputs.snmp.table]]
    oid = "1.3.6.1.4.1.9.9.63.1.3.8.4"
    name = "cvCallVolPeerTable"

    [[inputs.snmp.table.field]]
      oid = "1.3.6.1.4.1.9.9.63.1.3.8.4.1.1"
      name = "cvCallVolPeerIncomingCalls"

    [[inputs.snmp.table.field]]
      oid = "1.3.6.1.4.1.9.9.63.1.3.8.4.1.2"
      name = "cvCallVolPeerOutgoingCalls"
```

With these can you show the output of
```
telegraf --config telegraf.conf --input-filter snmp --test
```

**Comment by abitrusty** [comment]
Please see below. Appreciate the time
![image](https://user-images.githubusercontent.com/54104859/79781737-dba8ce80-8335-11ea-900e-96e854194277.png)

**Comment by danielnelson** [comment]
Let's add `index_as_tag = true` to the table, this way you should have an additional tag on each metric:
```toml
[[inputs.snmp]]
  ## ... snip ...

  [[inputs.snmp.table]]
    oid = "CISCO-VOICE-DIAL-CONTROL-MIB::cvCallVolPeerTable"
    index_as_tag = true
```

We can use that index in a processor to enhance the metric with the customer name:
```toml
[[processors.enum]]
  namepass = ["cvCallVolPeerTable"]
  [[processors.enum.mapping]]
    tag = "index"
    dest = "customer"
    default = "Unknown"
    [processors.enum.mapping.value_mappings]
      1 = "Customer 1"
      2 = "Customer 2"
```

To answer some of the questions brought up by @abitrusty above, you can put this processor configuration in the same file or any *.conf file in the `--config-directory` if you are using one.  I suggest keeping this close by to the input plugin since they go together logically.

It will run on every metric produced by all input plugins, which is why we add the `namepass` option to restrict it to only the needed metrics.  The processors run after the inputs and before the outputs.

One gotcha is that currently they are not run with `--test`, so to try it out I suggest temporarily switching your influxdb output for a file output writing to stdout:
```toml
[[outputs.file]]
  files = ["stdout"]
```

**Comment by abitrusty** [comment]
Hi - defintely looks promising and thanks for the info. I've spent some time on this and I'm not seeing any info at all so must be missing something simple

Have this test.conf file in /etc/telegraf/telegraf.d  but I do not see any reference to this agent IP in the metrics.out file at all & no metrics getting popualted to Influx for this agent

If I run a telegraf test on this I get
![image](https://user-images.githubusercontent.com/54104859/79849450-6501e480-83ba-11ea-8eac-c58eec7c5977.png)



[[inputs.snmp]]
 agents = [ "X.X.X.X" ]
 version = 2
 community = "XXXXX"
 interval = "60s"
 timeout = "10s"
 retries = 3

[[inputs.snmp.field]]
    name="hostname"
    oid = "SNMPv2-MIB::sysName.0"
    is_tag = true

[[inputs.snmp.field]]
    name = "uptime"
    oid = "DISMAN-EVENT-MIB::sysUpTimeInstance"

[[inputs.snmp.table]]
    oid = "CISCO-VOICE-DIAL-CONTROL-MIB::cvCallVolPeerTable"
    index_as_tag = true

[[processors.enum]]
  namepass = ["cvCallVolPeerTable"]
  [[processors.enum.mapping]]
    tag = "index"
    dest = "customer"
    default = "Unknown"
    [processors.enum.mapping.value_mappings]
      100 = "Customer 1"
      200 = "Customer 2"

[[outputs.file]]
  files = ["stdout", "/tmp/metrics.out"]


Only thing I see in the telegraf journalctl logs is:

Apr 21 10:27:00 p21graphs telegraf[26423]: Cannot find module (CISCO-VOICE-DIAL-CONTROL-MIB): At line 1 in (none)
Apr 21 10:27:00 p21graphs telegraf[26423]: CISCO-VOICE-DIAL-CONTROL-MIB::cvCallVolPeerTable: Unknown Object Identifier

I have tested snmptranslate & snmptable commands using the same MIB and it gets the info required. I have copied the Cisco MIBs into most places I see on the forums that require it on Ubuntu 18.04

**Comment by abitrusty** [comment]
It's working!! Daniel you are a legend!! Really appreciate it

The config I have above is good. I gave up earlier this morning but I just rebooted my telegraf server there an hour ago for another reason and then just noticed a new measurement called "CvCallVoIPeerTable" ...which has tags of "index" being my dialpeers and also "customer" - exactly what I need. 
Hope this helps someone else

![image](https://user-images.githubusercontent.com/54104859/79888530-b929ba80-83f4-11ea-9a03-bb8e098e7f1d.png)

**Comment by abitrusty** [comment]
It's working!! Daniel you are a legend!! Really appreciate it

The config I have above is good. I gave up earlier this morning but I just rebooted my telegraf server there an hour ago for another reason and then just noticed a new measurement called "CvCallVoIPeerTable" ...which has tags of "index" being my dialpeers and also "customer" - exactly what I need. 
Hope this helps someone else

![image](https://user-images.githubusercontent.com/54104859/79888530-b929ba80-83f4-11ea-9a03-bb8e098e7f1d.png)

**Comment by danielnelson** [comment]
Awesome, glad I could help.
